python/ray/serve/handle.py

```diff
@@ -61,46 +65,64 @@ def _create_or_get_global_asyncio_event_loop_in_thread():
 
 
 @dataclass(frozen=True)
-class _HandleOptions:
-    """Options for each ServeHandle instance.
+class _InitHandleOptionsBase:
+    """Init options for each ServeHandle instance.
+
+    These fields can be set by calling `.init()` on a handle before
+    sending the first request.
+    """
+
+    _prefer_local_routing: bool = False
+    _source: DeploymentHandleSource = DeploymentHandleSource.UNKNOWN
+
+
+@dataclass(frozen=True)
+class _InitHandleOptions(_InitHandleOptionsBase):
+    @classmethod
+    def create(cls, **kwargs) -> "_InitHandleOptions":
+        for k in list(kwargs.keys()):
+            if kwargs[k] == DEFAULT.VALUE:
+                # Use default value
+                del kwargs[k]
+
+        # Detect replica source for handles
+        if (
+            "_source" not in kwargs
+            and ray.serve.context._get_internal_replica_context() is not None
+        ):
+            kwargs["_source"] = DeploymentHandleSource.REPLICA
+
+        return cls(**kwargs)
+
+
+@dataclass(frozen=True)
+class _DynamicHandleOptionsBase:
+    """Dynamic options for each ServeHandle instance.
 
     These fields can be changed by calling `.options()` on a handle.
     """
 
     method_name: str = "__call__"
     multiplexed_model_id: str = ""
     stream: bool = False
-    _prefer_local_routing: bool = False
     _request_protocol: str = RequestProtocol.UNDEFINED
-    _source: DeploymentHandleSource = DeploymentHandleSource.UNKNOWN
 
-    def copy_and_update(
-        self,
-        method_name: Union[str, DEFAULT] = DEFAULT.VALUE,
-        multiplexed_model_id: Union[str, DEFAULT] = DEFAULT.VALUE,
-        stream: Union[bool, DEFAULT] = DEFAULT.VALUE,
-        _prefer_local_routing: Union[bool, DEFAULT] = DEFAULT.VALUE,
-        _request_protocol: Union[str, DEFAULT] = DEFAULT.VALUE,
-        _source: Union[DeploymentHandleSource, DEFAULT] = DEFAULT.VALUE,
-    ) -> "_HandleOptions":
-        return _HandleOptions(
-            method_name=(
-                self.method_name if method_name == DEFAULT.VALUE else method_name
-            ),
-            multiplexed_model_id=(
-                self.multiplexed_model_id
-                if multiplexed_model_id == DEFAULT.VALUE
-                else multiplexed_model_id
-            ),
-            stream=self.stream if stream == DEFAULT.VALUE else stream,
-            _prefer_local_routing=self._prefer_local_routing
-            if _prefer_local_routing == DEFAULT.VALUE
-            else _prefer_local_routing,
-            _request_protocol=self._request_protocol
-            if _request_protocol == DEFAULT.VALUE
-            else _request_protocol,
-            _source=self._source if _source == DEFAULT.VALUE else _source,
-        )
+    def copy_and_update(self, **kwargs) -> "_DynamicHandleOptionsBase":
+        pass
```

Comment:
is this an abstract base method?

Reply:
yes but not necessary, method seems general enough for now. I've put the implementation into the base class.